src/mac.cpp

```diff
@@ -0,0 +1,124 @@
+/*
+ * #%L
+ * %%
+ * Copyright (C) 2020 BMW Car IT GmbH
+ * %%
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * #L%
+ */
+#include "mococrw/mac.h"
+#include "mococrw/error.h"
+#include "mococrw/openssl_wrap.h"
+
+#include <openssl/hmac.h>
+#include <openssl/crypto.h>
+
+namespace mococrw
+{
+
+MessageAuthenticationCode::~MessageAuthenticationCode() = default;
+
+class HMAC::Impl
+{
+public:
+    Impl(openssl::DigestTypes hashFunction, const std::vector<uint8_t> key)
+    {
+        const EVP_MD* digestFn = _getMDPtrFromDigestType(hashFunction);
+
+        if (key.empty()) {
+            throw MoCOCrWException("Key for HMAC is empty.");
+        }
+
+        _ctx = openssl::_HMAC_CTX_new();
+        openssl::_HMAC_Init_ex(_ctx.get(), key, digestFn, NULL);
+    }
+
+    ~Impl() = default;
+
+    Impl(Impl&&) = default;
+
+    void update(const std::vector<uint8_t> &message)
+    {
+        if (_isFinished) {
+            throw MoCOCrWException("update() can't be called after finish()");
+        }
+        openssl::_HMAC_Update(_ctx.get(), message);
+    }
+
+    std::vector<uint8_t> finish()
+    {
+        if (_isFinished) {
+            throw MoCOCrWException("finish() can't be called twice.");
+        }
+
+        /* Currently SHA512 has the greatest hash size */
+        _result.resize(EVP_MAX_MD_SIZE);
+
+        unsigned int tagLength;
+        openssl::_HMAC_Final(_ctx.get(), _result, &tagLength);
+
+        _result.resize(tagLength);
+        _isFinished = true;
+
+        return _result;
+    }
+
+    void verify(const std::vector<uint8_t> &hmacValue)
+    {
+        if (!_isFinished) {
+            finish();
+        }
+
+        const size_t m = (hmacValue.size() < _result.size() ? hmacValue.size() : _result.size());
+        if (!!CRYPTO_memcmp(hmacValue.data(), _result.data(), m)) {
```

Comment:
Also taken from here:
https://wiki.openssl.org/index.php/EVP_Signing_and_Verifying#Calculating_HMAC
As I don't know if this triggers a compiler non-optimisation I have copied it. 

What do  you think?
Shall I remove it?

Reply:
I think the double-`!!` is in the OpenSSL code to ensure that it either return 0 or 1, regardless of whether `CRYPTO_memcmp` returns any value other than those two. Since we're not storing the result of the `CRYPTO_memcmp` in an `int`, we won't need it. Please remove it.

Reply:
Done